novatel_gps_driver/src/nodelets/wrapper/novatel_gps_nodelet_wrapper_worker.cpp

```diff
@@ -0,0 +1,113 @@
+{
+
+  NovatelGPSNodeletWrapperWorker::NovatelGPSNodeletWrapperWorker(WorkerConfig config, boost::shared_ptr<GPSDriverWrapper> publisher) : 
+    config_(config), publisher_(publisher)
+  {
+    status_.imu = true;
+    status_.gnss = true;
+    status_.name = publisher_->getQualifiedName();
+  }
+
+  void NovatelGPSNodeletWrapperWorker::imuCallback(const sensor_msgs::ImuConstPtr& msg) {
+    last_imu_msg_ = ros::Time::now();
+  }
+
+  void NovatelGPSNodeletWrapperWorker::inspvaxCallback(const novatel_gps_msgs::InspvaxConstPtr& msg) {
+
+    last_gnss_msg_ = ros::Time::now();
+
+    gps_common::GPSFix fix_msg;
+    // NOTE: At the moment CARMA's hardware interfaces do not use GPS status information
+    // Therefore that information is not currently populated in this message
+    // Convert position
+    fix_msg.header = msg->header;
+    fix_msg.latitude = msg->latitude;
+    fix_msg.longitude = msg->longitude;
+    fix_msg.altitude = msg->altitude; 
+    // Set position covariance
+    fix_msg.position_covariance_type = gps_common::GPSFix::COVARIANCE_TYPE_DIAGONAL_KNOWN;
+    fix_msg.position_covariance[0] = msg->longitude_std * msg->longitude_std;
+    fix_msg.position_covariance[4] = msg->latitude_std * msg->latitude_std;
+    fix_msg.position_covariance[8] = msg->altitude_std * msg->altitude_std;
+    
+    // Convert orientation
+    // NOTE: It is unclear in the message spec what dip represents so it will be ignored
+    fix_msg.track = msg->azimuth;
+    fix_msg.pitch = msg->pitch;
+    fix_msg.roll = msg->roll;
+
+    // GPSFix messages request uncertainty reported with 95% confidence intervale. That is 2 times the standard deviation
+    fix_msg.err_track = 2.0 * msg->azimuth_std;
+    fix_msg.err_pitch = 2.0 * msg->pitch_std;
+    fix_msg.err_roll = 2.0 * msg->roll_std;
+
+    // Convert Velocity
+    std::vector<double> components, variances;
+    components.push_back(msg->north_velocity);
+    variances.push_back(2.0 * msg->north_velocity_std);
+
+    components.push_back(msg->east_velocity);
+    variances.push_back(2.0 * msg->east_velocity_std);
+
+    std::tuple<double,double> speed_tuple = uncertainty_tools::computeVectorMagnitudeAndUncertainty(components, variances);
+    fix_msg.speed = std::get<0>(speed_tuple);
+    fix_msg.climb = msg->up_velocity;
+
+    fix_msg.err_climb = 2.0 * msg->up_velocity_std;
+    fix_msg.err_speed = std::get<1>(speed_tuple);
+
+    publisher_->publishFixFused(fix_msg);
+  }
+
+  void NovatelGPSNodeletWrapperWorker::statusTimerCallback(const ros::TimerEvent& event) {
+    ros::Time now = ros::Time::now();
+    switch(status_.status) {
+      case cav_msgs::DriverStatus::OFF:
+        if (last_gnss_msg_ != ros::Time(0) && last_imu_msg_ != ros::Time(0)) {
+          status_.status = cav_msgs::DriverStatus::OPERATIONAL;
+        }
+        break;
+      case cav_msgs::DriverStatus::OPERATIONAL:
+        if (now - last_gnss_msg_ > config_.gnss_timeout || now - last_imu_msg_ > config_.imu_timeout) {
```

Comment:
This check looks like the same as the one for DEGRADED state?

Reply:
This is the check to switch to FAULT state so it is the same when in the OPERATIONAL or DEGRADED state

Reply:
I see. Because we have no way to enter DEGRADED state, I was confused a little bit.